### PR TITLE
[WIP][SPARK-19937] Collect metrics of block sizes when shuffle.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

Metrics of blocks sizes(when shuffle) should be collected for later analysis. This is helpful for analysis when skew situations or OOM happens(though maxBytesInFlight is set).
This is a preparation for SPARK-19659.

## How was this patch tested?
Unit test in `HistoryServerSuite` and `JsonProtocolSuite`.

### CODE REVIEW

**Comment by jinxing64** [comment]
@squito 
Would you mind help comment on this when have time ? :)

File: core/src/main/scala/org/apache/spark/executor/ShuffleWriteMetrics.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/ShuffleWriteMetrics.scala]: the convention inside spark is to rename the java box classes w/ prefix"J", eg. `JLong`: https://demo.fluentcode.com/source/spark/master/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala?squery=JLong#L20

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: in scala, `if( isInstanceOf) { asInstanceOf}` can be replaced by pattern matching:

```scala
mapStatus match {
  case hc: HighlyCompressedMapStatus => writeMetrics.setAverageBlockSize(hc.getAvgSize());
  case _ =>  //no-op
}
```

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala]: scala convention -- simple getters don't have parens, `def getAvgSize: Long = avgSize`

File: core/src/main/scala/org/apache/spark/executor/ShuffleReadMetrics.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/ShuffleReadMetrics.scala]: since you are not using this yet, it shouldn't be part of this pr (I realize you may be using this as part of other changes in your own internal testing)

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: don't you want the condition reversed? `partitionLengths(i)` is the true size, so an underestimate is if the mapStatus says the size is smaller.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: another metric that may be nice to capture here is *maximum* underestimate -- `(0 until partitionLengths).map { i => partitionLengths(i) - mapStatus.getSizeForBlock(i) }.max`.  In fact, that alone might be enough to discover cases where the reduce side will OOM because of this underestimate.

**Review by squito — CHANGES_REQUESTED** [review]
Hi @jinxing64 , I left a few comments on the code for style, but my higher level concern is -- do you have any evidence that these metrics are actually useful in some scenarios?

Broadly speaking, I think they make sense.  However, there is some overhead to introducing metrics like these (even just for code maintenance), and they are exposed as a public api, so we should add them carefully.  Long ago, I was adding metrics very randomly to spark, but I think we need to be more careful now.

It would be great if you could try this out on some workloads and see if you are actually getting useful information.  Eg., on a highly skewed sort, maybe once w/ 100 partitions and once w/ 2001 partitions.  Or on whatever real workloads you have.

(I'd like to help more with this, but honestly I won't be able to for a while, so this will require more work on your part for now.)

**Comment by jinxing64** [comment]
@squito 
Thanks a lot for your comments and I will think and do the test carefully :)

**Comment by squito** [comment]
@jinxing64 do you mind closing this pr for now (or marking as [WIP] at least)?

**Comment by jinxing64** [comment]
@squito oh, I feel sorry if this is disturbing. I will mark it as wip.

**Comment by squito** [comment]
no worries, I'm just not sure when to look again, with all the notifications from your commits. Committers tend to think that something is ready to review if its passing tests, so its helpful to add those labels if its not the case.

**Comment by jinxing64** [comment]
You are so kind person.

**Comment by jinxing64** [comment]
@squito
Thanks a lot for taking time looking into this pr.
I updated the pr. Currently just add two metrics: a) the total size of underestimated blocks size, b) the size of blocks shuffled to memory.
For a), executor use `maxBytesInFlight` to control the speed of shuffle-read. I agree with your comment `another metric that may be nice to capture here is maximum underestimate`. But think about this scenario: the maximum is small, but thousands of blocks are underestimated, thus `maxBytesInFlight` cannot help avoid the OOM during shuffle-read. That's why I proposed to track the metrics of total size of underestimated blocks size;
For b), currently all data are shuffled-read to memory. If we add the feature of shuffling to disk when memory shortage, we need to evaluate the performance. I think another two metrics need to be taken into account: the size of blocks shuffled to disk(to be added in another pr) and task's running time(already exist). The more data shuffled to memory, the better performance; The shorter time cost, the better performance.

I also added some log for debug in `ShuffleWriter`, including the num of underestimated blocks and the size distribution.

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java
Comment (mridulm) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: This will essentially be sum of every block above average size - how is this supposed to be leveraged ?
For example:
1, 2, 3, 4, 5, 6 => 15
1, 2, 3, 4, 5, 10 => 15
(This ended up being a degenerate example - but in general, I am curious what the value is for this metric).

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java
Comment (mridulm) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: We need to handle case of mapStatus not being HighlyCompressedMapStatus also.

File: core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java
Comment (mridulm) [inline core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java]: This computation seems repeated - we should refactor it out into a method of its own and not duplicate it across classes.

File: core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleWriter.scala]: Isn;t this not similar to what is in core/src/main/java/org/apache/spark/shuffle/sort/UnsafeShuffleWriter.java, etc above ? Or is it different ?
The code looked same, but written differently (and more expensive here).

File: core/src/main/scala/org/apache/spark/executor/ShuffleReadMetrics.scala
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/ShuffleReadMetrics.scala]: The way it seems to be coded up, this will end up being everything fetched from shuffle - and we can already infer it : remote bytes read + local bytes read.
Or did I miss something here ?

**Comment by jinxing64** [comment]
@mridulm 
Thanks a lot for taking time looking into this and thanks for comments :)
1) I changed the size of underestimated blocks to be `partitionLengths.filter(_ > hc.getAvgSize).map(_ - hc.getAvgSize).sum`
2) I added a method `genBlocksDistributionStr` and call it from `ShuffleWriters`, thus avoid duplicate codes

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java
Comment (jinxing64) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: In `CompressedMapStatus`, the blocks sizes are accurate, so I might hesitate to add that log.

**Comment by mridulm** [comment]
@jinxing64 
If the intent behind these metrics is to help with SPARK-19659, it would be good to either add it as part of SPARK-19659 or subsequently (once the feature is merged).
This ensures that the metrics added are actually relevant to the existing spark core, and not a future expected evolution of the code - for example, the review of SPARK-19659 might significantly change its design/implementation : making some of these either irrelevant or require other more informative metrics to be introduced.

I am unclear about the intention btw - do you expect shuffle reads to be informed by metrics from the mapper side ? I probably got that wrong.

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java
Comment (mridulm) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: The value is not accurate - it is a 1og 1.1 'compression' which converts the long size to a byte : and caps the value at 255.

So there are two errors introduced; it over-estimates the actual block size when compressed value < 255 [1] (which is something this PR currently ignores), when block size goes above 34k mb or so, it under estimates the block size (which is higher than what spark currently supports due to 2G limitation).


[1] I did not realize it always over-estimates; if the current PR is targetting only blocks which are under estimated; I would agree that not handling `CompressedMapStatus` for time being might be ok - though would be good to add a comment to that effect on 'why' we dont need to handle it.

**Comment by jinxing64** [comment]
@mridulm 
Sorry for late reply.  I opened the pr for SPARK-19659(https://github.com/apache/spark/pull/16989) and make these two PRs independent. Basically this pr is is to evaluate the performance(blocks are shuffled to disk) and stability(size in `MapStatus` is inaccurate and OOM can happen) of the implementation proposed in SPARK-19659.
I'd be so thankful if you have time to  comment on these two PRs.

**Comment by jinxing64** [comment]
@mridulm @squito 
Thanks a lot for taking time review this pr.
I will close it for now and make another one if there is progress.

**Pull request closed without merging**
